File: TwoPointer/49.group-anagrams.py
from collections import defaultdict
from typing import List


# @leet start
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list)
        for s in strs:
            k = "".join(sorted(s))
            res[k].append(s)

        return list(res.values())


# @leet end

File: TwoPointer/49.group-anagrams.go
package leetcode

// @leet start
func groupAnagrams(strs []string) [][]string {
	var (
		freq  = map[[128]rune][]string{}
		parse = func(s string) (res [128]rune) {
			for _, r := range s {
				res[r]++
			}
			return res
		}
	)
	for _, s := range strs {
		key := parse(s)
		freq[key] = append(freq[key], s)
	}
	ret := make([][]string, 0, len(freq))
	for _, v := range freq {
		ret = append(ret, v)
	}

	return ret
}

// @leet end

